platforms/core-configuration/declarative-dsl-core/src/main/resources/org/gradle/internal/declarativedsl/parsing/grammar.bnf

```diff
@@ -44,22 +44,23 @@
 
 root ::= statements
 
-statement ::= (assignment | expression)
-statements ::= (statement semis?)*
+statement ::= (assignment | block | expression)
+statements ::= (statement semis*)*
 
-assignment ::= directlyAssignableExpression EQ primaryExpression {pin=2}
-expression ::= simpleIdentifier postfixUnarySuffixes
+assignment ::= directlyAssignableExpression EQ primaryExpression (navigationSuffix* valueArguments)? {pin=2}
+block ::= simpleIdentifier valueArguments? annotatedLambda
+expression ::= primaryExpression postfixUnarySuffixes
 
-semis ::= SEMI+
+semis ::= SEMI
 
 directlyAssignableExpression ::= simpleIdentifier navigationSuffix?
 
-postfixUnarySuffixes ::= (valueArguments? annotatedLambda) | (navigationSuffix* valueArguments)
+postfixUnarySuffixes ::= (navigationSuffix+ valueArguments) | (valueArguments (navigationSuffix valueArguments)*)
 
 annotatedLambda ::= LB statements RB {pin=1}
 
 valueArguments ::= LP (valueArgument (COMA valueArgument)* COMA?)? RP {pin=2}
-valueArgument ::= (simpleIdentifier EQ)? primaryExpression postfixUnarySuffixes?
+valueArgument ::= (simpleIdentifier EQ)? (expression | primaryExpression)
```

Comment:
`expression | primaryExpression` is where keeping the original naming from the Kotlin grammar starts looking really confusing to me. At this point, I'd prefer introducing our own meaningful names for the rules.

Reply:
Well, we could rename:

- `primaryExpression` --> `literalOrIdentifier`
- `expression` --> `functionCalls`, because that's what it is, simple, complex and chained function calls:
  - f()
  - f(x=y)
  - f.g.h.i.j.k(test)
  - f(1, x, "s", g())
  - f(a = b, c = d)
  - id("plugin").version("7")
  - f(1).g(2).h(3)

Would that work for you?

Reply:
`literalOrIdentifier` 👍 
For `functionCalls`, I'd prefer a singular noun, so that it is clear that all the calls can only be in one chain (compare to `statements` where those are unrelated statements).